### PR TITLE
feat: add deployment URL

### PR DESCRIPTION
**Pull request by denysvitali**
This PR, which has https://github.com/hashicorp/waypoint-plugin-sdk/pull/27 as a pre-requisite, adds support for Deployment URLs.

With this PR, it is now possible to show in the CLI the deployment URLs provided by the plugins, without the need of Horizon (URL Service).

**Warning:** The PR contains a `go.mod` entry that points to our fork since the PR in [https://github.com/hashicorp/waypoint-plugin-sdk](waypoint-plugin-sdk) hasn't been merged yet.

### CODE REVIEW

File: builtin/aws/lambda/deployment.go
Comment (mitchellh) [inline builtin/aws/lambda/deployment.go:8]: I think this should be optional so these no-ops don't have to be implemented (this is more of SDK feedback, but repeating it here) and also so we don't break any previous plugins.

File: builtin/k8s/deployment.go
Comment (mitchellh) [inline builtin/k8s/deployment.go:49]: Maybe I'm being dense, but where are all these being populated?

File: internal/cli/deployment_create.go
Comment (mitchellh) [inline internal/cli/deployment_create.go:103]: I think we need to add a `releaseUrl != ""` check in the output below so that we don't output an empty release URL.

File: internal/core/app_deploy.go
Comment (mitchellh) [inline internal/core/app_deploy.go:140]: Is there a reason behind this change related to adding a deployment URL?

If there is not a reason, OR its not related to deployment URLs, let's table this for another issue to discuss. This feels unrelated to this PR.

File: internal/server/server.go
Comment (mitchellh) [inline internal/server/server.go:16]: This appears unrelated to this PR. Please remove.

**Comment by krantzinator** [comment]
Thanks for taking a swing at this! I'm going to leave a couple of small comments here while the sdk side is sorted out.

File: internal/core/app_deploy.go
Comment (krantzinator) [inline internal/core/app_deploy.go:140]: I think this change was auto-generated somehow, as I don't see a functional reason for it. Can you revert this?

File: internal/server/proto/server.proto
Comment (mitchellh) [inline internal/server/proto/server.proto:1858]: This is fine, but as a bigger comment I think we should rethink our data model slightly around URLs. Right now it is unclear what the difference is between `deploy.Url` and `deploy.Preload.Url` and both MIGHT be empty or both MIGHT NOT be empty or ONE may be empty, etc. Actionably, I think we should:

* Have more detailed comments.
* Have a canonical URL field that has SOME URL if one exists (may still be empty if none exist)
* Clearly delineate between URL service URL and Platform URL

File: internal/core/app_deploy.go
Comment (denysvitali) [reply internal/core/app_deploy.go:140]: This change was made to be in line with Release. In there the `Component` is exported, I'm not sure why - but I wanted to make it consistent :P

**Review by mitchellh — CHANGES_REQUESTED** [review]
The idea here is A+, no complaints, I think we should support this. I think the implementation side needs a bit of work. There's one big comment in there around data modeling.

I think you're using a different version of Mockery which caused changes in all the generated mocks. Please use the pinned version we have in our Makefile. Thanks!

File: internal/server/gen/mocks/waypoint_server.go
Comment (krantzinator) [inline internal/server/gen/mocks/waypoint_server.go:1]: Could you put the mockery version change in a different PR, or maybe at least a different commit? There are a couple of places this is pinned, so it would be nice to capture the whole change at once.

File: internal/server/server.go
Comment (denysvitali) [reply internal/server/server.go:16]: Will do :)

File: internal/server/proto/server.proto
Comment (denysvitali) [reply internal/server/proto/server.proto:1858]: I didn't found any documentation around `deploy.Preload.Url`. I think it's supposed to be an Horizon pre-defined URL before the application is even deployed, is that correct?

File: internal/server/gen/mocks/waypoint_server.go
Comment (krantzinator) [reply internal/server/gen/mocks/waypoint_server.go:1]: Can disregard this comment in favor of Mitchell's review :)

File: builtin/aws/lambda/deployment.go
Comment (denysvitali) [reply builtin/aws/lambda/deployment.go:8]: I'm not sure how we can make it optional to be honest, what would you suggest in this case?

Comment (mitchellh) [reply builtin/aws/lambda/deployment.go:8]: No problem! I think a separate interface would be the way to go (in the SDK):

```go
type DeploymentWithURL interface {
  Deployment

  URL() string
}
```

File: builtin/k8s/deployment.go
Comment (denysvitali) [reply builtin/k8s/deployment.go:49]: It's not being populated in fact. This is just an empty URL at the moment, waiting to be implemented:
https://github.com/hashicorp/waypoint/pull/1387/files/1dcacbb891bf4f181adfd3c952a96373073648be#diff-1b0cb4abccfe10f5d0dcb1e343487a8da3a3da4c7b06dd4e86be69b5456681f1R10

File: internal/core/app_deploy.go
Comment (denysvitali) [reply internal/core/app_deploy.go:140]: Unrelated, sorry. It was just a change to be consistent with:
https://github.com/swisscom/waypoint/blob/feature/deployment-url/internal/core/app_release.go#L152

File: internal/server/gen/mocks/waypoint_server.go
Comment (denysvitali) [reply internal/server/gen/mocks/waypoint_server.go:1]: You're right Rae, I will remove this from the PR since it's out of scope

File: internal/server/proto/server.proto
Comment (mitchellh) [reply internal/server/proto/server.proto:1858]: Yes, because our data model only understands URL service URLs currently. :) That's the big conceptual shift in this PR I think is that we're now saying deployments may have a platform-provided URL. But the data model doesn't quite reflect that and I think it should.

Comment (denysvitali) [reply internal/server/proto/server.proto:1858]: How would you like to approach this?

Do you have any proposed change?

**Comment by denysvitali** [comment]
Since the CI checks are failing (due to the fact that https://github.com/hashicorp/waypoint-plugin-sdk/pull/27 is still not merged), I temporarily added a replacement in `go.mod`, so that the tests are passing.

Please, do not merge if the SDK change isn't merged first.

**Comment by denysvitali** [comment]
@mitchellh / @briancain can you please re-start the integration test? I can't do it from my side

**Comment by krantzinator** [comment]
Reran the failed workflow

**Comment by denysvitali** [comment]
Ping @krantzinator @mitchellh or @briancain , do you know what's wrong with the `vendor` folder check?

Btw, rebased on top of main and good to go :)

**Comment by krantzinator** [comment]
@denysvitali looks like you need to run a `go mod tidy`

**Comment by denysvitali** [comment]
Thanks, tests are OK now ;)

File: .changelog/1387.txt
Comment (mitchellh) [inline .changelog/1387.txt:2]: We don't add internal changes to the changelog, recommendation below:

```suggestion
core: platform plugins may now advertise deployment-specific URLs
```

Comment (denysvitali) [reply .changelog/1387.txt]: I have no clue on why it is not working for me (using Nix as well). A double hypen fixes the issue for me, but as we already discussed this should be addressed in another PR. Sorry, it might have slipped into my commit.

File: internal/server/server.go
Comment (briancain) [reply internal/server/server.go:16]: I believe Mockery version 2.x+ uses double hyphens but we're on an older version so we use single.

Comment (denysvitali) [reply internal/server/server.go:16]: But does the nix environment use a newer one?  Because in my nix shell, the default (single hypen) wasn't working IIRC

Comment (briancain) [reply internal/server/server.go:16]: I'm not sure, I don't use it 😄 But when I was first getting started I just installed the latest Mockery and this was an issue I ran into, rolling back to `1.1.2` fixed it for me.

https://github.com/hashicorp/waypoint/blob/7d6229c3167209b80459e67bdc669c32465aceb0/tools/tools.go#L22-L23

Comment (briancain) [reply internal/server/server.go:16]: Looks like it should technically be installing the right version.... https://github.com/hashicorp/waypoint/blob/7d6229c3167209b80459e67bdc669c32465aceb0/shell.nix#L61-L62

Comment (denysvitali) [reply internal/server/server.go:16]: Really, really strange. I'll have to double check my environment since shell.nix reports the 1.x version too 😮

Comment (denysvitali) [reply internal/server/server.go]: Reverted the change, sorry!

**Comment by denysvitali** [comment]
Uh. Just noticed that after some changes it is not working anymore (deployment.Url is not filled). Let me re-work on this.

**Comment by denysvitali** [comment]
Okay, so let me unpack the new changes.

### `Deployment.Url` is now used only internally

Since we want to avoid having redundant fields, `Deployment.Url` will now be used only internally. The CLI and the UI will only consider `Deployment.Preload.DeployUrl`

We use `setDeploymentUrlIfNeeded` to write `d.Preload.DeployUrl` from `d.Url` if the latter exists.
This only happens when the client requests some `Deployment` resources (e.g: `GetDeployment` and `ListDeployments`), exactly like the URL service invocation works.

### The CLI now displays deployment URLs too

Kind of self-explanatory. Here is an example:

```
$ waypoint deployment list -workspace=develop
     | ID |   PLATFORM   |  DETAILS   |    STARTED     |   COMPLETED    |                   URL                   |     HEALTH
-----+----+--------------+------------+----------------+----------------+-----------------------------------------+-----------------
  ✖  | 12 | cloudfoundry | artifact:3 | 9 seconds ago  | 8 seconds ago  |                                         | Unknown status
  🚀 | 11 | cloudfoundry | artifact:3 | 14 minutes ago | 14 minutes ago | xy-api-v11.internal.swisscom.ch | Unknown status
  ✔  | 10 | cloudfoundry | artifact:3 | 50 minutes ago | 50 minutes ago | xy-api-v10.internal.swisscom.ch | Unknown status
```

### The UI now shows Deployment URLs too

Previously a button "Visit Deployment" was shown - now the full URL is shown, exactly like for the Releases

Screenshot 2021-06-09 at 20 58 58


### The Deployment view now shows `v10` instead of `#10`

Before my changes, this part was inconsistent: v0 was displayed in the general view but `#10` was shown when the deployment was clicked.

Screenshot 2021-06-09 at 20 59 40

**Comment by denysvitali** [comment]
Friendly ping :)

@briancain

File: internal/cli/deployment_list.go
Comment (briancain) [inline internal/cli/deployment_list.go:284]: I think given this bug: https://github.com/hashicorp/waypoint/issues/1608
We should only be showing the deploy URL with the `verbse` flag. Otherwise I think deployment list might be pretty difficult to format and look nice until that is fixed.

**Review by briancain — CHANGES_REQUESTED** [review]
Thanks for the ping @denysvitali - sorry this took a while to get a final review!

I had a small comment about the deploy URLs in `waypoint deployment list`, and I think it would be best to separate the UI changes in a follow-up PR. Then we can get the core functionality in with this at least and get a proper UI review on the follow-up PR.

FYI I am around today (pacific time in the US) but I am on vacation next week 🏖️ so someone else from the team can help out and review next week on any updates or on the UI PR 😄

**Comment by denysvitali** [comment]
I'm solving #1608 by adding a new URL flag (to display it), because one is probably not interested in seeing all the additional details but might want to see the deployment URL:

```
$ waypoint deployment list -workspace=develop
     | ID |   PLATFORM   |   DETAILS   |  STARTED   | COMPLETED  | HEALTH  
-----+----+--------------+-------------+------------+------------+---------
  🚀 | 64 | cloudfoundry | artifact:34 | 1 week ago | 1 week ago | n/a     
  ✔  | 63 | cloudfoundry | artifact:33 | 1 week ago | 1 week ago | n/a    
```


```
$ waypoint deployment list -workspace=develop -V
     | ID |   PLATFORM   |              DETAILS               |  STARTED   | COMPLETED  | HEALTH  
-----+----+--------------+------------------------------------+------------+------------+---------
  🚀 | 64 | cloudfoundry | artifact:34                        | 1 week ago | 1 week ago | n/a     
     |    |              | artifact.product:the-product    
     |    |              | build.product:the-product       
     |    |              | deployment.product:the-product  
  ✔  | 63 | cloudfoundry | artifact:33                        | 1 week ago | 1 week ago | n/a     
     |    |              | artifact.product:the-product    
     |    |              | build.product:the-product       
     |    |              | deployment.product:the-product  
```


```
$ waypoint deployment list -workspace=develop -u
     | ID |   PLATFORM   |   DETAILS   |  STARTED   | COMPLETED  | HEALTH |                   URL                    
-----+----+--------------+-------------+------------+------------+--------+------------------------------------------
  🚀 | 64 | cloudfoundry | artifact:34 | 1 week ago | 1 week ago | n/a    | product-1-c1f94b83.xyz.swisscom.ch  
  ✔  | 63 | cloudfoundry | artifact:33 | 1 week ago | 1 week ago | n/a    | product-1-1c414a16.xyz.swisscom.ch  
```

**Comment by denysvitali** [comment]
Test are failing because of a network issue:

> Error making request.
> Error: connect ETIMEDOUT 140.82.113.4:443
>    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1144:16)

**Comment by krantzinator** [comment]
If a plugin hasn't implemented deployment URLs, what is the expected behavior? I'm not seeing anything listed in this output for e.g. the nomad plugin. I would expect it would show the default at that point?
```
$ waypoint deployment list -u
     | ID | PLATFORM |               DETAILS               |   STARTED    |  COMPLETED   |      HEALTH      | URL  
-----+----+----------+-------------------------------------+--------------+--------------+------------------+------
  🚀 |  1 | nomad    | image:8cc9f30                       | 1 minute ago | 1 minute ago | ✔ - 1 minute ago |      
```

**Comment by denysvitali** [comment]
> If a plugin hasn't implemented deployment URLs, what is the expected behavior?

This is a very good question! We could eventually fall back to the `b.Preload.Url` (which should be the Hashicorp Horizon URL), if available, or return N/A if none of them are available.

What do you think?

**Comment by krantzinator** [comment]
> > If a plugin hasn't implemented deployment URLs, what is the expected behavior?
> 
> This is a very good question! We could eventually fall back to the `b.Preload.Url` (which should be the Hashicorp Horizon URL), if available, or return N/A if none of them are available.
> 
> What do you think?

That makes sense to me!

**Comment by denysvitali** [comment]
@krantzinator Fixed as agreed

File: internal/cli/deployment_list.go
Comment (krantzinator) [inline internal/cli/deployment_list.go:298]: quick bug fix and a couple of syntactical preferences to match the repo's style
```suggestion
				url := "n/a"
				if b.Url != "" {
					url = b.Url
				} else if b.Preload.DeployUrl != "" {
					url = b.Preload.DeployUrl
				}
				columns = append(columns, url)
```

**Comment by krantzinator** [comment]
Looking good! Thanks for sticking with this! I gave one last suggestion, and then it looks like we need to do a rebase and regenerate the protobufs. Let me know if you'd like me to do that. 🙌🏻

File: internal/cli/deployment_list.go
Comment (denysvitali) [reply internal/cli/deployment_list.go]: Sounds good! I tend not to use `url` because it conflicts with the `net.url` package, but since we're not using it here all good!

**Comment by denysvitali** [comment]
> Let me know if you'd like me to do that. 🙌🏻

Yes, please do it!

**Comment by krantzinator** [comment]
@denysvitali can you give me access to the forked repo so I can edit in this PR rather than making a new branch?

**Comment by denysvitali** [comment]
@krantzinator : No clue on why GitHub doesn't let me give you guys permissions over the branch itself, so I added you as a writer to our repository :)
The invite is still pending though

**Comment by krantzinator** [comment]
@denysvitali I got it! Just waiting on the docker server build to finish re-caching the new go mod stuff so I can give it a quick spin before merging :)

**Review by krantzinator — APPROVED** [review]

**Pull request merged**
